### PR TITLE
add ownership rest apis

### PR DESCRIPTION
**Pull request by shichanglin5**
# What problem are we solving?

[Controlling object ownership](https://docs.aws.amazon.com/AmazonS3/latest/userguide/about-object-ownership.html)

# How are we solving the problem?

# How is the PR tested?

# Checks
- [ ] I have added unit tests if possible.
- [ ] I will add related wiki document changes and link to this PR after merging.

### CODE REVIEW

File: weed/s3api/filer_util.go [hunk 1]
```diff
@@ -91,6 +91,22 @@ func (s3a *S3ApiServer) getEntry(parentDirectoryPath, entryName string) (entry *
 	return filer_pb.GetEntry(s3a, fullPath)
 }
 
+func (s3a *S3ApiServer) updateEntry(parentDirectoryPath string, newEntry *filer_pb.Entry) (entry *filer_pb.Entry, err error) {
```
Comment (chrislusf) [inline weed/s3api/filer_util.go]: what is the first return value?

File: weed/s3api/s3api_acp.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package s3api
+
+import (
+	"github.com/seaweedfs/seaweedfs/weed/s3api/s3_constants"
+	"github.com/seaweedfs/seaweedfs/weed/s3api/s3err"
+	"net/http"
+)
+
+func ensureAccountId(r *http.Request) string {
```
Comment (chrislusf) [inline weed/s3api/s3api_acp.go]: rename to getAccountId()?

File: weed/s3api/s3api_acp.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	id := r.Header.Get(s3_constants.AmzAccountId)
+	if len(id) == 0 {
+		return AccountAnonymous.Id
+	} else {
+		return id
+	}
+}
+
+func (s3a *S3ApiServer) OwnershipAccess(r *http.Request, bucket string) s3err.ErrorCode {
```
Comment (chrislusf) [inline weed/s3api/s3api_acp.go]: rename to checkAccessByOwnership()?

File: weed/s3api/s3api_bucket_handlers.go [hunk 4]
```diff
@@ -343,3 +345,161 @@ func (s3a *S3ApiServer) GetBucketLocationHandler(w http.ResponseWriter, r *http.
...
+		return
+	}
+
+	if r.Body == nil || r.Body == http.NoBody {
+		s3err.WriteErrorResponse(w, r, s3err.ErrInvalidRequest)
+		return
+	}
+
+	var v s3.OwnershipControls
+	defer func(Body io.ReadCloser) {
```
Comment (chrislusf) [inline weed/s3api/s3api_bucket_handlers.go]: defer util.CloseRequest(r)

File: weed/s3api/s3api_bucket_handlers.go [hunk 5]
```diff
@@ -343,3 +345,161 @@ func (s3a *S3ApiServer) GetBucketLocationHandler(w http.ResponseWriter, r *http.
...
+	if err != nil {
+		if err == filer_pb.ErrNotFound {
+			s3err.WriteErrorResponse(w, r, s3err.ErrNoSuchBucket)
+			return
+		}
+		s3err.WriteErrorResponse(w, r, s3err.ErrInternalError)
+		return
+	}
+
+	oldOwnership, ok := bucketEntry.Extended[s3_constants.ExtOwnershipKey]
```
Comment (chrislusf) [inline weed/s3api/s3api_bucket_handlers.go:401]: bucketEntry.Extended can be nil.

File: weed/s3api/auth_credentials.go [hunk 6]
```diff
@@ -259,6 +286,7 @@ func (iam *IdentityAccessManagement) authRequest(r *http.Request, action Action)
 		return identity, s3err.ErrAccessDenied
 	}
 
+	r.Header.Set(s3_constants.AmzAccountId, identity.AccountId)
```
Comment (chrislusf) [inline weed/s3api/auth_credentials.go]: is this necessary for anonymous mode?

File: weed/s3api/s3api_bucket_handlers.go (hunk 5 above)
Comment (shichanglin5) [reply weed/s3api/s3api_bucket_handlers.go:401]: A nil map can be read directly, but writing will cause NPE, so it doesn't matter here

File: weed/s3api/auth_credentials.go (hunk 6 above)
Comment (shichanglin5) [reply weed/s3api/auth_credentials.go]: Do you mean anonymous access? There is no need to set it for anonymous access. When the request header reads `X-Amz-Account` is empty, the id of `AccountAnonymous` will be used as accountId
```go
func ensureAccountId(r *http.Request) string {
	id := r.Header.Get(s3_constants.AmzAccountId)
	if len(id) == 0 {
		return AccountAnonymous.Id
	} else {
		return id
	}
}
```

File: weed/s3api/s3api_bucket_handlers.go (hunk 5 above)
Comment (famosss) [reply weed/s3api/s3api_bucket_handlers.go:401]: > bucketEntry.Extended can be nil.

Tested, no exception will be thrown here.

File: weed/s3api/auth_credentials.go (hunk 6 above)
Comment (famosss) [reply weed/s3api/auth_credentials.go]: > is this necessary for anonymous mode?

There are still several `pr` operation objects that need to use this header. The anonymous user is the anonymous accountId, which should not matter.

Comment (chrislusf) [reply weed/s3api/auth_credentials.go]: how about only add this header when the user is not anonymous?

File: weed/s3api/auth_credentials.go [hunk 7]
```diff
@@ -259,6 +287,9 @@ func (iam *IdentityAccessManagement) authRequest(r *http.Request, action Action)
 		return identity, s3err.ErrAccessDenied
 	}
 
+	if !identity.isAnonymous() {
+		r.Header.Set(s3_constants.AmzAccountId, identity.AccountId)
+	}
 	return identity, s3err.ErrNone
```
Comment (shichanglin5) [inline weed/s3api/auth_credentials.go:293]: > how about only add this header when the user is not anonymous?

is this ok @chrislusf

**Pull request merged**